improver/calibration/reliability_calibration.py

```diff
@@ -488,3 +491,193 @@ def process(self, cubes, coordinates=None):
+        Warns:
+            UserWarning: If the probabilities must be sorted to reinstate
+                         expected monotonicity following calibration.
+        """
+        threshold_dim, = cube.coord_dims(self.threshold_coord)
+        thresholding = self.threshold_coord.attributes.get(
+            "spp__relative_to_threshold", None)
+
+        if thresholding is None:
+            msg = ('Cube threshold coordinate does not define whether '
+                   'thresholding is above or below the defined thresholds.')
+            raise ValueError(msg)
+
+        if (thresholding == 'above' and not
+                (np.diff(cube.data, axis=threshold_dim) <= 0).all()):
+            msg = ('Exceedance probabilities are not decreasing monotonically '
+                   'as the threshold values increase. Forced back into order.')
+            warnings.warn(msg)
+            cube.data = np.sort(cube.data, axis=threshold_dim)[::-1]
+
+        if (thresholding == 'below' and not
+                (np.diff(cube.data, axis=threshold_dim) >= 0).all()):
+            msg = ('Below threshold probabilities are not increasing '
+                   'monotonically as the threshold values increase. Forced '
+                   'back into order.')
+            warnings.warn(msg)
+            cube.data = np.sort(cube.data, axis=threshold_dim)
+
+    def _calculate_reliability_probabilities(self, reliability_table):
+        """
+        Calculates forecast probabilities and observation frequencies from the
+        reliability table. Where the forecast count is zero, the returned
+        arrays both contain values of zero.
+
+        Args:
+            reliability_table (iris.cube.Cube):
+                A reliability table for a single threshold from which to
+                calculate the forecast probabilities and observation
+                frequencies.
+        Returns:
+            (tuple): tuple containing:
+                **forecast_probability** (numpy.array):
+                    Forecast probabilities calculated by dividing the sum of
+                    forecast probabilities by the forecast count.
+                **observation_frequency** (numpy.array):
+                    Observation frequency calculated by dividing the
+                    observation count by the forecast count.
+        """
+        observation_count = reliability_table.extract(
+            iris.Constraint(table_row_name='observation_count')).data
+        forecast_count = reliability_table.extract(
+            iris.Constraint(table_row_name='forecast_count')).data
+        forecast_probability_sum = reliability_table.extract(
+            iris.Constraint(
+                table_row_name='sum_of_forecast_probabilities')).data
+
+        # In some bins have insufficient counts, return None to avoid applying
+        # calibration.
+        valid_bins = np.where(forecast_count >= self.minimum_forecast_count)
+        if valid_bins[0].size != forecast_count.size:
```

Comment:
That's not quite what this is saying. The function np.where returns a tuple of 1-d arrays, each of the same length by necessity, as all combined they give the index of each position that we want to use. As such, I cannot use a ```valid_bins, =``` construct as unless I have a 1-d input array (which is pretty much never the case here), this will cause an error (too many values to unpack). I need to take any of the arrays returned within the tuple and check the size, and zero is guaranteed to always be there.